File: Frontend/lib/models/recipe_model.dart
class RecipeModel {
  final String? id;
  final String? name;
  final String? description;
  final List<String>? ingredients;
  final List<String>? preparation;
  final Map<String, dynamic>? nutrition;
  final int? servingMin;
  final int? servingMax;
  final int? cookTime;
  final int? difficulty;

  RecipeModel({
    this.id,
    this.name,
    this.description,
    this.ingredients,
    this.preparation,
    this.nutrition,
    this.servingMin,
    this.servingMax,
    this.cookTime,
    this.difficulty
  });

  factory RecipeModel.fromJson(Map<String, dynamic> json) {
    return RecipeModel(
      id: json['_id'],
      name: json['name'],
      description: json['description'],
      ingredients: List<String>.from(json['ingredients'] ?? []),
      preparation: List<String>.from(json['preparation'] ?? []),
      nutrition: Map<String, dynamic>.from(json['nutrition'] ?? []),
      servingMin: json['serving']['min'] as int?,
      servingMax: json['serving']['max'] as int?,
      cookTime: json['cookTime'] as int?,
      difficulty: json['difficulty'] as int?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'description': description,
      'ingredients': ingredients,
      'preparation': preparation,
      'nutrition': nutrition,
      'serving': {'min': servingMin, 'max': servingMax},
      'cookTime': cookTime,
      'difficulty': difficulty,
    };
  }
}

File: Backend/models/recipe.model.js
const mongoose = require("mongoose");

const RecipeSchema = new mongoose.Schema({
    // 'name' is a required field of type String
    // represents the name of the recipe
    name: {
      type: String,
      required: true,
    },
    // 'description' is a required field of type String
    // represents the description of the recipe
    description: {
      type: String,
      required: true,
    },
    // 'ingredients' is a required field of type String
    // represents the ingredients in a recipe
    ingredients: [
      {
        type: String,
        required: true,
      },
    ],
    // 'preparation' is a required field of type String
    // represents the steps for the preparation of a recipe
    preparation: [
      {
        type: String,
        required: true,
      },
    ],
    // 'serving' is a required field of type Number
    // represents the number of people who can feed of a single plate
    serving: {
      min: {
        type: Number,
        required: false,
      },
      max: {
        type: Number,
        required: true,
      },
    },
    // 'cookTime' is not a required field of type Number
    // represents the time that takes to cook a recipe
    cookTime: {
      type: Number,
      required: false,
    },
    // 'name' is not a required field of type Number
    // represents between 1-3 the dificulty of a recipe
    difficulty: {
      type: Number,
      required: false,
    },
    // 'nutrition' is a required field of type Number
    // represents the prots, carbs and fibre of a recipe
    nutrition: {
      protein: {
        type: Number,
        required: true,
      },
      carbs: {
        type: Number,
        required: true,
      },
      fibre: {
        type: Number,
        required: true,
      },
    },
  },
  { collection: 'recipes' },
  { timestamps: true }
);

module.exports = mongoose.model("recipes", RecipeSchema);
